src/mlpack/methods/rbm/spike_slab_layer_impl.hpp

```diff
@@ -0,0 +1,149 @@
+/**
+ * @file spike_slab_hidden_impl.hpp
+ * @author Kris Singh
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RBM_SPIKE_SLAB_SPIKE_SLAB_LAYER_IMPL_HPP
+#define MLPACK_METHODS_RBM_SPIKE_SLAB_SPIKE_SLAB_LAYER_IMPL_HPP
+// In case it hasn't yet been included.
+#include "spike_slab_layer.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/dists/gaussian_distribution.hpp>
+using namespace mlpack::distribution;
+
+namespace mlpack {
+namespace rbm { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+SpikeSlabLayer<InputDataType, OutputDataType>::SpikeSlabLayer(
+    const size_t inSize,
+    const size_t outSize,
+    const size_t poolSize,
+    const double radius,
+    const bool typeVisible):
+    inSize(inSize),
+    outSize(outSize),
+    poolSize(poolSize),
+    radius(radius),
+    typeVisible(typeVisible)
+{
+  weights.set_size(
+      inSize * ((outSize * inSize) + (poolSize * poolSize) + 1) + outSize,
+      outSize * outSize);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void SpikeSlabLayer<InputDataType, OutputDataType>::Reset()
+{
+  if (typeVisible)
+  {
+    visibleSize = inSize;
+    hiddenSize = outSize;
+  }
+  else
+  {
+    visibleSize = outSize;
+    hiddenSize = inSize;
+  }
+  // Weight shape = k * d * n
+  weight = arma::cube(weights.memptr(), poolSize,
+      visibleSize, hiddenSize, false,
```

Comment:
Well i agree with that. But i think the weight. shape in ann are generally outsize * insize. That's why i followed the convention here.